library/VTKExtensions/Rendering/vtkF3DRenderer.cxx

```diff
@@ -1340,6 +1341,39 @@ void vtkF3DRenderer::ShowEdge(bool show)
   }
 }
 
+//----------------------------------------------------------------------------
+void vtkF3DRenderer::SetUseOrthographicProjection(bool use)
+{
+  vtkCamera* camera = GetActiveCamera();
+  const double angle = vtkMath::RadiansFromDegrees(camera->GetViewAngle());
+  const double* position = camera->GetPosition();
+  const double* focal = camera->GetFocalPoint();
+
+  if (this->UseOrthographicProjection != use)
+  {
+    this->UseOrthographicProjection = use;
+
+    const double distance = std::sqrt(vtkMath::Distance2BetweenPoints(position, focal));
+    const double parallelScale = distance * tan(angle / 2);
+    camera->SetParallelScale(parallelScale);
+  }
+  else
+  {
+    this->UseOrthographicProjection = !use;
+
+    const double distance = camera->GetParallelScale() / tan(angle / 2);
+    double direction[3];
+    vtkMath::Subtract(position, focal, direction);
+    vtkMath::Normalize(direction);
+    vtkMath::MultiplyScalar(direction, distance);
+    double newPosition[3];
+    vtkMath::Add(focal, direction, newPosition);
+    camera->SetPosition(newPosition);
+  }
+  camera->SetParallelProjection(use);
+  this->ResetCameraClippingRange();
```

Comment:
I won't claim the logic here doesn't work but it feels more complicated than it has to be.
Can't it go something like this?
```c++
SetUseOrthographicProjection(bool use){

  // if the internal state is already the same as the target state there's nothing to do
  if(this->UseOrthographicProjection == use) 
    return;

  // grab camera variables
  vtkCamera* camera = GetActiveCamera();

  if (use) // switching to ortho
  { 
     // set new parallel scale from current distance
  }
  else // switching to perspective
  { 
    // set new distance from current parallel scale
  }

  camera->SetParallelProjection(use); // actually do the switch
  this->UseOrthographicProjection = use; // update internal state

  this->ResetCameraClippingRange(); // needed to avoid rendering issues
}
```

Reply:
It is rather complicated. Are you saying that it would be better to add 2 helper functions and call those from the 'if then else'?

Reply:
no need for helper functions, it's just about the checking/setting of the internal state.

In your current code you do:
```cpp
  if (this->UseOrthographicProjection != use)
  {
    this->UseOrthographicProjection = use;
    // set new parallel scale from current distance
  }
  else
  {
    this->UseOrthographicProjection = !use;
    // set new distance from current parallel scale
  }
  camera->SetParallelProjection(use);
```
and I think that is confusing 2 things:
- checking the internal state: we _want_ to compare the current state and the target state at the top of the function so we can bail early if there's nothing to do
- checking the target state: we _need_ to do this to know which mode we're switching to and adjust the camera accordingly

Reply:
I fully agree with @snoyer analysis here. :)